README
Signal Server 1.3.3 changelog 

v1.3.3 - 04 Nov 2012
Air planning:
Increased maximum Tx height to 60,000 (m/f)
Increased maximum Rx height to 60,000 (m/f)

v1.3.2 - 04 Oct 2012
Re-instated grey scale option to allow for terrain background.
Off by default. Enable with switch -t

v1.3 - 03 Jan 2012
Added user defined clutter layers from SPLAT! (-udt switch)

v1.2 - 31 Dec 2011
Max colours increased from 32 to 128
Radius value fixed at metric
Metric / Imperial conversion bug fixed

v1.1 - 08 Dec 2011
Max Tx altitude increased to 20,000(m) for high altitude aircraft. "Can you 'ere me now?"
Added 2 new options -tercon, -terdic for *custom* dielectric values and ground conductivity. -te terrain option remains. Use with care!
New Earth Dielectric range (Permittivity): 80 to 0.1
New Conductivity range (Siemens/m): 0.01 to 0.000001

v1.0 - 19 November 2011
SS released. 







CHANGELOG
Signal Server 1.3.8 changelog 

v1.3.8 - 16 Jan 2014
Added Free Space Path Loss model (with optional diffraction)

v1.3.7 - 30 Dec 2013
Added propagation model option (-pm)
Added HATA urban/suburban/open models (150-1500MHz) 
Added COST231-Hata (urban) model (1500-2000MHz) 
Added custom Knife Edge Diffraction option (-ked) to enhance new models
Removed unused variables 

v1.3.6 - 12 Aug 2013
Added LOS model for up to 100GHz 
Added 1 arc second (3600 pixels/degree) support (-res 3600)

v1.3.5 - 07 Jul 2013
Reduced maxpages to 9
Added memset() to clear DEM before use

v1.3.4 - 16 May 2013
High resolution SRTM1 1-arc second DEM support added.
Use -res 3600 and ensure .sdf files are produced with srtm2sdf-hd

v1.3.3 - 04 Nov 2012
Air planning:
Increased maximum Tx height to 60,000 (m/f)
Increased maximum Rx height to 60,000 (m/f)

v1.3.2 - 04 Oct 2012
Re-instated grey scale option to allow for terrain background.
Off by default. Enable with switch -t

v1.3 - 03 Jan 2012
Added user defined clutter layers from SPLAT! (-udt switch)

v1.2 - 31 Dec 2011
Max colours increased from 32 to 128
Radius value fixed at metric
Metric / Imperial conversion bug fixed

v1.1 - 08 Dec 2011
Max Tx altitude increased to 20,000(m) for high altitude aircraft. "Can you 'ere me now?"
Added 2 new options -tercon, -terdic for *custom* dielectric values and ground conductivity. -te terrain option remains. Use with care!
New Earth Dielectric range (Permittivity): 80 to 0.1
New Conductivity range (Siemens/m): 0.01 to 0.000001

v1.0 - 19 November 2011
SS released. 







cost.cpp
/*****************************************************************************
*  COST231-HATA MODEL for Signal Server by Alex Farrant                         *  
*  30 December 2013										                     *
*  This program is free software; you can redistribute it and/or modify it   *
*  under the terms of the GNU General Public License as published by the     *
*  Free Software Foundation; either version 2 of the License or any later    *
*  version.								   								     *
* 									    									 *
*  This program is distributed in the hope that it will useful, but WITHOUT  *
*  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or     *
*  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License     *
*  for more details.													     *
*									     									 */

#include <stdlib.h>
#include <math.h>
#include <iostream>

using namespace std;

double CostHataLinkdB(float f,float h_B, float h_M, float d){
/*
COST HATA URBAN model
Frequency 1500 to 2000MHz
h_B = Base station height 30 to 200m
h_M = Mobile station height 1 to 10m
Distance 1-20km
*/

	int C = 0; // 0dB for suburban
	
	float lh_M = log10(11.75*h_M);
	float C_H = 3.2*lh_M*lh_M-4.97;
	
	float logf = log10(f);
	
    double dbloss = 46.3 + (33.9 * logf) - (13.82 * log10(h_B)) - C_H + (44.9 - 6.55 * log10(h_B)) * log10(d) + C;  


	return dbloss;
}


fspl.cpp
/*****************************************************************************
*  ITU-R P.525 Free Space Path Loss model for Signal Server by Alex Farrant  *  
*  15 January 2014										                     *
*  This program is free software; you can redistribute it and/or modify it   *
*  under the terms of the GNU General Public License as published by the     *
*  Free Software Foundation; either version 2 of the License or any later    *
*  version.								   								     *
* 									    									 *
*  This program is distributed in the hope that it will useful, but WITHOUT  *
*  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or     *
*  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License     *
*  for more details.													     *
*									     									 */

#include <stdlib.h>
#include <math.h>
#include <iostream>

using namespace std;

double FsplLinkdB(float f, float d){
/*
Free Space Path Loss model
Frequency: Any
Distance: Any 
*/
	//MHz to GHz
	f = f / 1000;

	double dbloss = (20 * log10(d)) + (20 * log10(f)) + 92.45;
   
	return dbloss;
}


hata.cpp
/*****************************************************************************
*  HATA MODEL for Signal Server by Alex Farrant                              *
*  30 December 2013								                             *
*  This program is free software; you can redistribute it and/or modify it   *
*  under the terms of the GNU General Public License as published by the     *
*  Free Software Foundation; either version 2 of the License or any later    *
*  version.								   								     *
* 									    									 *
*  This program is distributed in the hope that it will useful, but WITHOUT  *
*  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or     *
*  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License     *
*  for more details.													     *
*									     									 */

#include <stdlib.h>
#include <math.h>
#include <iostream>
#include <stdio.h>
#include <string>

using namespace std;

#define PI 3.14159265

/* Acute Angle from Rx point to an obstacle of height (opp) and distance (adj) */ 
double incidenceAngle(double opp, double adj){
return atan2(opp,adj) * 180 / PI;
}

/* 
Knife edge diffraction:
This custom function adds to the overall path loss based upon the obstacle 
angle (from receive point) and the frequency. Loss increases with angle and frequency.
This is not a recognised formula like Huygens, rather it is a 
compromise for increased speed which adds a realistic diffraction effect.
*/
double ked(double freq, double elev[], double rxh, double dkm){
double obh,obd,rxobaoi=0,d;

obh=0;
obd=0;

dkm=dkm*1000; // KM to metres

	for(int n=2;n<(dkm/elev[1]);n++){
	
	d = (n-2)*elev[1]; // no of points * delta = km
	
	    //Find dip(s)
		if(elev[n] > 0 && elev[n]<(obh+10)){
		
		// Angle from Rx point to obstacle
		rxobaoi = incidenceAngle((obh-(elev[n]+rxh)),d-obd);
		} else{
		// Line of sight or higher
		rxobaoi=0;
		}
	
		//note the highest point
		if(elev[n]>obh){
		obh=elev[n];
		obd=d; 
		}
		
	}
	
if(rxobaoi >= 0){
return (rxobaoi * 3) / (300/freq); // Exaggerate diffraction angle and divide by wavelength (m)
}else{
return 0;
}

}

double HataLinkdB(float f,float h_B, float h_M, float d, int mode){
/*
HATA URBAN model for cellular planning
Frequency (MHz) 150 to 1500MHz
Base station height 30-200m
Mobile station height 1-10m
Distance 1-20km

mode 1 = URBAN
mode 2 = SUBURBAN
mode 3 = OPEN
*/

	float lh_M = log10(11.75*h_M);
	float C_H = 3.2*lh_M*lh_M-4.97;
	
	float logf = log10(f);
	
	float L_u = 69.55 + 26.16*logf - 13.82*log10(h_B) - C_H + (44.9 - 6.55*log10(h_B))*log10(d);

	if(!mode || mode==1){
	return L_u; //URBAN
	}
	
	if(mode==2){ //SUBURBAN
	float logf_28 = log10(f/28);
	return L_u - 2*logf_28*logf_28 - 5.4;
	}

	if(mode==3){ //OPEN
	return L_u - 4.78*logf*logf + 18.33*logf - 40.94;
	}

	return 0;
}